eks: Implement deploy script and guide for EKS (#78)
testground/infra: k8s/eks/README.md

Links have been included for each software requirement. Listed guides are for example purposes only.**

**1. docker**

Official installation guide:

https://docs.docker.com/engine/install/ubuntu/

Installation on amazon linux:

```
sudo yum install docker -y
sudo systemctl enable docker.service
sudo systemctl start docker.service
sudo usermod -aG docker ec2-user
newgrp docker
```

**2. helm**

Official installation guide:

https://helm.sh/docs/intro/install/

```
curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
chmod 700 get_helm.sh
./get_helm.sh
```

**3. kubectl**

Official installation guide:

https://kubernetes.io/docs/tasks/tools/

Latest version:

```
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
kubectl version --client
```

Specific version (required when your cluster is not compatible with the latest version):

```
curl -LO https://dl.k8s.io/release/v1.22.0/bin/linux/amd64/kubectl
sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
kubectl version --client
```

> **4. AWS CLI (v2) with AWS credentials and ECR login**

Reviewer: Do we use ECR login during the installation script?

Author: We do not, that's why this is in the README:
```
It is assumed that you already have the following:
- A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script```

Reviewer: But we can remove `and ECR login` part, right?

Author: Well when it comes to the bastion, I believe we can remove it. I need to get back to you on this one, need to test if it is safe to remove.

Author: ECR login part has been removed from the README.

This guide will help you setup an EKS (Elastic Kubernetes Service) cluster, and run testground plans on it.

## Requirements

It is assumed that you already have the following:

> - A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script, interact with the cluster, and run testground plans

Author: AWS docs are not complete regarding this; I have created a step-by-step instruction which will be shared with you.

> - A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script, interact with the cluster, and run testground plans

Reviewer: Can you publish it as a draft so that we can already have some insight?

> - A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script, interact with the cluster, and run testground plans

Author: ```
1. Create a keypair

AWS Console -> EC2 -> Network & Security -> Key Pairs -> Create key pair (upper right corner)
Input key name
Pick RSA
Pick .pem as you will be creating a linux bastion host
Create the key and download the private key. You will use it to connect to the bastion

2. Create an instance

AWS Console -> EC2 -> Instances -> Launch instances (upper right corner)
Add an instance name
Pick Quick Start and Amazon Linux
Instance type t2.micro
Pick your key pair
Pick a VPC - either an existing one (for example the default VPC), or create a new VPC
Pick Create a new Security Group and allow SSH traffic only from your public IP, not 0.0.0.0/0
Make sure that 'Auto-assign public IP' is set to 'Enable'
Root volume (storage) can be 30GB gp2

3. Once the instance shows as 'Running' with both Status Checks as '2/2 checks passed', you may attempt to ssh into the bastion host
on macOS you may use:

ssh -i key-file ec2-user@IP-Address (obtain the public IP address of the bastion host through the AWS Console)
key-file is the private key file you downloaded in step 1.

4. Update the instance and install necessary software listed in the README

Since using Amazon Linux, it should be:

sudo yum update

And then you may install the rest of the software listed in README. 

Note: Amazon Linux is a RHEL/CentOS derivative

5. After everything is installed, pull the infra and testground repos, and continue following the README
```

> - A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script, interact with the cluster, and run testground plans

Reviewer: I think it'd be better with a link to https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/get-set-up-for-amazon-ec2.html

> - A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script, interact with the cluster, and run testground plans

Author: Yeah sure, we can add that as well.

> - A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script, interact with the cluster, and run testground plans

Author: Added the link to the README.

- The laptop or bastion host has the following software installed:

> **Note: the software and installation guides have been tested on amazon linux. Please verify which OS are you using and follow the official guides accordingly.

Author: Why create a custom AMI, and why to use packer, could you please elaborate this? Nothing against the tool, just asking why suddenly introducing another tool to the mix.

> **Note: the software and installation guides have been tested on amazon linux. Please verify which OS are you using and follow the official guides accordingly.

Reviewer: It's strictly better than a set of instructions that humans have to perform manually. This is error prone and likely will get outdated pretty quickly. Well written packer build script is more reproducible.

Anyway, I'm just asking to insert a TODO here. By no means I'm asking you to develop it right now.

> **Note: the software and installation guides have been tested on amazon linux. Please verify which OS are you using and follow the official guides accordingly.

Author: Ah ok, got it. Will be added.

> **Note: the software and installation guides have been tested on amazon linux. Please verify which OS are you using and follow the official guides accordingly.

Author: Added a TODO for this effort.

Links have been included for each software requirement. Listed guides are for example purposes only.**

**1. docker**

Official installation guide:

https://docs.docker.com/engine/install/ubuntu/

Installation on amazon linux:

```
sudo yum install docker -y
sudo systemctl enable docker.service
sudo systemctl start docker.service
sudo usermod -aG docker ec2-user
newgrp docker

> ```

Author: While I agree with you, these comments were added with the main purpose of aiding you with first time setup and not wasting time searching for correct commands as the official docs are sometimes not updated. 
These are actually tested commands and could save you time for initial testing.

> ```

Reviewer: I'd think docker official docs would be getting more frequent updates than the testground ones. That's why I'd like to see links to official docs instead of specific instructions that work on your machine.

> ```

Author: And that is why we have included both links to the offical docs as first lines for each dependency, and added these notes to make it easier for you or any other user. 
Please read the README file again.

**2. helm**

Official installation guide:

https://helm.sh/docs/intro/install/

```
curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
chmod 700 get_helm.sh

> ./get_helm.sh

Author: We could of course, and some some third-party dependencies we have included this as well. 
But have in mind that this needs to be tested and updated frequently, as it is easy to reach the software rot stage.

> ./get_helm.sh

Reviewer: In the current form, this will become untested as soon as https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3 starts returning a new version.

Please use pinned versions where possible.

> ./get_helm.sh

Author: Ok, we will use pinned versions where possible.

> ./get_helm.sh

Author: Specific versions of software used during writing of this have been added to the README where applicable.

```

**3. kubectl**

Official installation guide:

https://kubernetes.io/docs/tasks/tools/

Latest version:

```
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
kubectl version --client
```

> Specific version (required when your cluster is not compatible with the latest version):

Author: We could of course, and some some third-party dependencies we have included this as well. 
But have in mind that this needs to be tested and updated frequently, as it is easy to reach the software rot stage.

> Specific version (required when your cluster is not compatible with the latest version):

Reviewer: What is software rot stage?

> Specific version (required when your cluster is not compatible with the latest version):

Author: If you pin a specific version of a software, and then do not maintain it or test it, it will soon rot and become unusable.

> Specific version (required when your cluster is not compatible with the latest version):

Reviewer 2: Stability & reproducibility is more important here, let's pin versions as requested by @galargh

https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html#cliv2-linux-install

Then configure the credentials:

https://docs.aws.amazon.com/cli/latest/userguide/getting-started-quickstart.html

Verify you can log into ECR using the set credentials:

```
aws ecr get-login-password --region $AWS_REGION | docker login --username AWS --password-stdin $ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com

# or in this case since the images are moved to a public repo:

aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/n6b0k8i7
```

**5. eksctl**

Official installation guide:

https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html

```
curl --silent --location "https://github.com/weaveworks/eksctl/releases/latest/download/eksctl_$(uname -s)_amd64.tar.gz" | tar xz -C /tmp
sudo mv /tmp/eksctl /usr/local/bin
eksctl version
```

## Step by step installation guide

1. Pull the infra repo from here:

git clone https://github.com/testground/infra.git

and the testground repo from here:

git clone https://github.com/testground/testground

> 2. You need to populate the `.env` file with the needed parameters before creating the cluster. It is located in `infra/k8s/eks/`

Author: Agree, we will have to outline the most important parameters.

> 3. Run the installation script `./testground_install.sh`; note that it requires the `.env` file to be populated in order to run properly. In case of any missing variables, the setup will fail and you will get an error message

Author: If you recall the Phase report that was sent at the beginning of September, it describes this exact effort:

```
We are considering two options to replace the current cluster deployment scripts (written in bash).

Eksctl is a CLI tool used to create and manage EKS clusters. It is simple to implement, and satisfies most demands that we have of the EKS cluster. However, it is not as customizable and maintainable as Terraform.

Terraform is more robust and customizable, but it takes more time to implement and test.

For the 1st new version of the deployment scripts, the team has elected to use eksctl, as a shell script. In the next phase, we will perform a cost-benefit analysis of using Terraform, and discuss this with other stakeholders and PL team members.
```

The script explained in short:

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Author: I think it more depends on the number of worker nodes. But if eksctl shows as completed, we should still expect some more time to pass before the nodes are running per the integrated AWS checks.

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Reviewer: So how does increasing number of worker nodes increase the cluster setup time?

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Author: I tested with 35 nodes in a NG, it is a maybe 1-2 minutes difference overall.
The main difference would be on the AWS side as the worker nodes need to boot up and pass some checks, as explained above.

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Reviewer: I just want a confirmation that at the end of the script we're not waiting to ensure that the setup is ready to use. Is that correct?

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Author: Please understand that some things are completely out of our reach. And that sometimes AWS takes more time to deploy nodes.
The script will finish its run and will tell us that the Cloudformation templates have deployed successfully. But AWS is ultimately running checks for deployed instances.
This is related to AWS and there is nothing we can do to ensure that.

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Reviewer: What nodes are you talking about here? Is there no API to check if those nodes are ready?

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Author: A function called `wait_for_alb_and_instances` has been implemented to solve this.

> - Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads

Reviewer: Great!

- All installation steps have been divided into functions, located in the `eks/bash/functions.sh` script
- The script calls functions one by one
- Output of every step is logged into a master log file that can be retrieved and reviewed, or even tailed in real time

> - It can also be used to uninstall a provisioned cluster; one of the first steps in the script is checking for existing resources

Author: A separate `testground_uninstall.sh` script has been created.

- The `functions.sh` file relies on the content of `eks/yaml` - all cluster resources are located inside this folder

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Author: By default, we are creating 2 worker nodes per NG. If you want to initially scale that, you need to edit the `functions.sh` on line 44 (the `make_cluster_config` function that creates the NG yaml file)
Also, scaling explanation has been added to the README since you left this comment.

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Reviewer: I can't see where we set 2 worker nodes per NG in the place you mentioned. When you mention a specific file/line, please add a permalink to that place.

Could we make that setting configurable through env?

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Author: So the `.env` file already has this capability, and it is called `DESIRED_CAPACIY_INFRA=2` and `DESIRED_CAPACIY_PLAN=2`, which is then used in the `make_cluster_config` function (listed entire function below so you can find it easily:

```
make_cluster_config(){
    cat <<EOT >> ./$CLUSTER_NAME.yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: $CLUSTER_NAME
  region: $REGION
managedNodeGroups:
  - name: ng-1-infra
    labels:
      "testground.node.role.infra": "true"
    instanceType: $INSTANCE_TYPE_INFRA
    ami: $AMI_INFRA
    desiredCapacity: $DESIRED_CAPACITY_INFRA
    volumeSize: $VOLUME_SIZE_INFRA
    privateNetworking: false
    availabilityZones: ["$AVAILABILITY_ZONE"]
    # ssh:
    #   allow: true
    #   # key needs to exist
    #   publicKeyPath: $SSH_PATH_INFRA
    iam:
      attachPolicyARNs:
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess
        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
        - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
   
    overrideBootstrapCommand: |
      #!/bin/bash
      /etc/eks/bootstrap.sh $CLUSTER_NAME \
        --kubelet-extra-args '--max-pods=58' \
        --use-max-pods false
  - name: ng-2-plan
    labels:
      "testground.node.role.plan": "true"
    instanceType: $INSTANCE_TYPE_PLAN 
    # Amazon EKS optimized Amazon Linux 2 v1.22 built on 08 Aug 2022
    ami: $AMI_PLAN
    desiredCapacity: $DESIRED_CAPACITY_PLAN
    volumeSize: $VOLUME_SIZE_PLAN
    privateNetworking: false
    availabilityZones: [$AVAILABILITY_ZONE]
    # ssh:
    #   allow: true
    #   publicKeyPath: $SSH_PATH_PLAN
    iam:
      attachPolicyARNs:
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess
        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
        - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
    preBootstrapCommands:
      - "echo -e 'fs.file-max=3178504\nnet.core.somaxconn = 131072\nnet.netfilter.nf_conntrack_max = 1048576\nnet.core.netdev_max_backlog = 524288\nnet.core.rmem_max = 16777216\nnet.core.wmem_max = 16777216\nnet.ipv4.tcp_rmem = 16384 131072 16777216\nnet.ipv4.tcp_wmem = 16384 131072 16777216\nnet.ipv4.tcp_mem = 262144 524288 1572864\nnet.ipv4.tcp_max_syn_backlog = 131072\nnet.ipv4.ip_local_port_range = 10000 65535\nnet.ipv4.tcp_tw_reuse = 1\nnet.ipv4.ip_forward = 1\nnet.ipv4.conf.all.rp_filter = 0\nnet.ipv4.neigh.default.gc_thresh2 = 4096\nnet.ipv4.neigh.default.gc_thresh3 = 32768' >> /tmp/999-testground.conf; sudo cp /tmp/999-testground.conf /etc/sysctl.d/999-testground.conf; sudo sysctl -p /etc/sysctl.d/999-testground.conf; echo -e '*  soft  nproc  131072\n*  hard  nproc  262144\n*  soft  nofile 131072\n*  hard  nofile 262144' >> /tmp/999-limits.conf; sudo cp /tmp/999-limits.conf /etc/security/limits.d/999-limits.conf"
    # 234 is the max number of pods for c5.4xlarge
    overrideBootstrapCommand: |
      #!/bin/bash
      /etc/eks/bootstrap.sh $CLUSTER_NAME \
       --kubelet-extra-args '--max-pods=234' \
       --use-max-pods false
EOT
}
```

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Reviewer: So we only have to modify env and `functions.sh` stays in tact, right? If that's the case, let's update the note in the README then.

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Author: Yes, exactly. It was intended to have only one config file that you need to edit, which is `.env`. Everything else runs with the parameters you specify in that file.
Makes things easier and simpler to have only one entrypoint to the script.

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Reviewer: But the README still says
```
- If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script
```

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Author: Yeah, that is a mistake and will be corrected. It should say `.env` instead of `functions.sh`

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Reviewer: Why don't we do it straight away? Is there any reason why we're putting changes like that off?
```suggestion
- If you need to scale worker nodegroups, modify `DESIRED_CAPACIY_INFRA` or `DESIRED_CAPACIY_PLAN` in the environment file (`.env`)
```

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Author: Because there are many more edits for the README file that I need to introduce, and I don't want to do it without checking some other things first.
Also, could you please edit your proposed change as there are 2 typos in it - should be `CAPACITY` instead of `CAPACIY`. 
Thanks.

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Reviewer: If you see something off there, just do a counter proposal. That's what this feature is for.

> - If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script

Reviewer: Gathering all the changes is one huge blob is what got us here in the first place. Doing more smaller changes reduces review burden massively.

- Once the cluster is created, eksctl will automatically switch the context to the new cluster. If you create another cluster, eksctl will once again switch the context to the newest cluster. If you wish to switch to another cluster, issue:

```
kubectl config get-contexts

# currently selected cluster will be mark with an asterisk (*)
# then switch contexts to the desired cluster

kubectl config use-context $USERNAME@$CLUSTER_NAME.$AWS_REGION.eksctl.io
```

> 4. Create the `.env.toml` file inside the `testground` folder and populate it (explained here as well https://github.com/testground/docs/blob/master/getting-started.md#configuration-envtoml):

Author: Agree, we will look into introducing that.

> 4. Create the `.env.toml` file inside the `testground` folder and populate it (explained here as well https://github.com/testground/docs/blob/master/getting-started.md#configuration-envtoml):

Author: The script now outputs the `.env.toml` once it finishes, so the users only need to copy and paste it in their `testground/.env.toml` file.
README updated with this change.

> **Note: The endpoint refers to the `testground-daemon` service, so depending on your setup, this could be, for example, a Load Balancer fronting the kubernetes cluster and forwarding proper requests to the `tg-daemon` service, or a simple port forward to your local workstation.**

Author: Could you please clarify this?

> **Note: The endpoint refers to the `testground-daemon` service, so depending on your setup, this could be, for example, a Load Balancer fronting the kubernetes cluster and forwarding proper requests to the `tg-daemon` service, or a simple port forward to your local workstation.**

Reviewer: When I look at the output of `testground_install` script, how do I find the value I should use for the `endpoint`? Could we mark it in a way to would make it obvious?

> **Note: The endpoint refers to the `testground-daemon` service, so depending on your setup, this could be, for example, a Load Balancer fronting the kubernetes cluster and forwarding proper requests to the `tg-daemon` service, or a simple port forward to your local workstation.**

Author: I believe the script doesn't output that in this iteration, but it can be added.
Right now it is explained in the README under step `6.` how to set up the LB.

> **Note: The endpoint refers to the `testground-daemon` service, so depending on your setup, this could be, for example, a Load Balancer fronting the kubernetes cluster and forwarding proper requests to the `tg-daemon` service, or a simple port forward to your local workstation.**

Reviewer: We should have that information readily available. If we could get https://github.com/testground/infra/pull/78#discussion_r984576809 though, it'd be perfect.

> **Note: The endpoint refers to the `testground-daemon` service, so depending on your setup, this could be, for example, a Load Balancer fronting the kubernetes cluster and forwarding proper requests to the `tg-daemon` service, or a simple port forward to your local workstation.**

Author: Yeah I agree, we will have to modify the script so it outputs this for you.
It is now missing because we added LB support late last week and didn't work on the script as well.

> **Note: The endpoint refers to the `testground-daemon` service, so depending on your setup, this could be, for example, a Load Balancer fronting the kubernetes cluster and forwarding proper requests to the `tg-daemon` service, or a simple port forward to your local workstation.**

Author: The script now outputs the .env.toml once it finishes, so the users only need to copy and paste it in their testground/.env.toml file.
It also includes the public DNS of the LB.

`.env.toml` example:

```
["aws"]
region = "$YOUR_REGION" # where the cluster will be created
[client] # Note: Only one endpoint can be active, please refer to steps 5. and 6. for clarification
endpoint = "http://localhost:8080" # if using port-forwarding

> endpoint = "http://acca15c6ah3eh45n68ad6052c20ba9ec-1569872837.eu-west-3.elb.amazonaws.com:80" # if using LB (default)

Author: Well for local testing I guess, it is all about preference.
It is not hardcoded to a LB, it is a random string that shows how a LB looks like so the users can know what to look for.

> endpoint = "http://acca15c6ah3eh45n68ad6052c20ba9ec-1569872837.eu-west-3.elb.amazonaws.com:80" # if using LB (default)

Reviewer: How about we make the instructions assume LB then and describe how to set up port forwarding separately in the "advanced" part of the documentation?

> endpoint = "http://acca15c6ah3eh45n68ad6052c20ba9ec-1569872837.eu-west-3.elb.amazonaws.com:80" # if using LB (default)

Author: This is exactly what was done on Friday. README was updated to reflect it.
But we can move the port forwarding section in another place; it was logical to me to keep it at the same place as the LB.

> endpoint = "http://acca15c6ah3eh45n68ad6052c20ba9ec-1569872837.eu-west-3.elb.amazonaws.com:80" # if using LB (default)

Reviewer: @laurentsenta What do you think?

> endpoint = "http://acca15c6ah3eh45n68ad6052c20ba9ec-1569872837.eu-west-3.elb.amazonaws.com:80" # if using LB (default)

Reviewer 2: > Well for local testing I guess, it is all about preference.

Is there such a thing as local testing when you work with an EKS cluster?

> How about we make the instructions assume LB then and describe how to set up port forwarding separately in the "advanced" part of the documentation?

That makes sense to me: the documentation will target a "regular" user starting an EKS cluster.
Port forwarding should target advanced users trying to work on EKS or Testground itself.

(waiting for the next review back and forth to re-review)

> endpoint = "http://acca15c6ah3eh45n68ad6052c20ba9ec-1569872837.eu-west-3.elb.amazonaws.com:80" # if using LB (default)

Author: Ok, agree. The README now reflects LB as the default case, while the port-forward has been moved to the `Advanced Setup` section of the README.

user = "myname"
```

> 5. Using the simple port-forward solution

Author: I would leave it there for the sake of user preference, and having seen how useful it can be as it is fairly simple to use.
And I believe that people could benefit from that information.

> 5. Using the simple port-forward solution

Reviewer: As in https://github.com/testground/infra/pull/78#discussion_r985626524, I'd suggest describing port-forwarding setup separately then. What I'd like us to achieve is a simple README that takes you through the most common path.

> 5. Using the simple port-forward solution

Author: As answered [here](https://github.com/testground/infra/pull/78#discussion_r985634903), we could move the port forward to a different section.

> 5. Using the simple port-forward solution

Author: Done, answered [here](https://github.com/testground/infra/pull/78#discussion_r989377865)

**NOTE: By default, the script will deploy the testground service with a LB in front. If you would like to use port-forward instead of the LB, you need to edit the `eks/yaml/tg-testground-daemon-service.yml` to look like this:**

```
apiVersion: v1
kind: Service
metadata:
  name: testground-daemon
    labels:
    app: testground-daemon
spec:
  ports:
  - port: 8042
    protocol: TCP
  selector:
    app: testground-daemon
```

Issue the following command in another terminal window and you should see something like this:

```
kubectl port-forward service/testground-daemon 8080:8042
```

where 8042 is the port on which the `tg-daemon` is listening, and 8080 is a port on your local workstation/bastion host.

```
Forwarding from 127.0.0.1:8080 -> 8042
Forwarding from [::1]:8080 -> 8042
```

This means you are successfully forwarding traffic to the testground daemon from your localhost; remember that you've specified this in the `.env.toml` config file.

> 6. Setup the Load Balancer

Author: We will look into adding this info.

> 6. Setup the Load Balancer

Author: Done, answered [here](https://github.com/testground/infra/pull/78#discussion_r989377749).

```
endpoint = "http://acca15c6ah3eh45n68ad6052c20ba9ec-1569872837.eu-west-3.elb.amazonaws.com:80"
```

You can now run testground tests without having to port-forward to your laptop/bastion host.

7. Now you are ready to run testground tests:

```
# ping-pong test
testground run single --plan network --testcase ping-pong --builder docker:go --runner cluster:k8s --instances 2

# storm test
testground run single --plan=benchmarks --testcase=storm --builder=docker:go --runner=cluster:k8s --instances=10

# You can also use the provided script `eks/bash/perf.sh` to run multiple storm tests:

./perf.sh <run_count> <plan> <case> <instance_count> <builder> <runner>

# e.g. to schedule 10 runs with 50 pods/instances each:

./perf.sh 10 benchmarks storm 50 docker:go cluster:k8s
```

8. After the test has finished, obtain results:

```
testground collect --runner=cluster:k8s $RUN_ID
```

where the $RUN_ID will be generated and shown in the test output. It looks like this `cccb96126un8ibqhe7p0`.

9. After you are done, you can delete the cluster since AWS resources cost money while running:

> Simply run the `testground_install.sh` script again and you will be prompted to remove the previously created resources, as the script will automatically detect them:

Author: A separate testground_uninstall.sh script has been created.

```
 "We found that you already have a cluster provisioned with this script"
 "Do you want to remove it? (y/n)?"
```

## Optional steps

By default, ssh access to the worker nodes is not enabled. In case you do need to access your worker nodes, do the following:

- Generate an ssh key on your laptop/bastion (which you use to create the cluster):

```
ssh-keygen -t rsa -b 4096
```

The keys will be saved in `~/.ssh/` or `/home/$USERNAME/.ssh`, as both `id_rsa` and `id_rsa.pub`.

> Then, in the installation script, edit the `make_cluster_config` function and uncomment the following:

Author: We will look into remodeling this part.
As for SSH access to already running cluster, we have used a daemonset for adding SSH keys to worker nodes during the testing phase, so we could clean it up and attach as an option here.

> Then, in the installation script, edit the `make_cluster_config` function and uncomment the following:

Author: ssh has been completely removed from the script and moved to the `Advanced Setup` section of the README.
Also added an explanation on setting up ssh to already running worker nodes.

```
    ssh:
      allow: true
      publicKeyPath: $SSH_PATH_INFRA  # note that this needs to reflect the path to your generated ssh key. Or you can set up a variable in the `.env` file and refer it here as an env var, like shown in the example
```

**Important note: when you allow ssh access like this, eksctl will create another AWS Security Group which allows ssh traffic (port 22) from `0.0.0.0/0` and `::/0`. This means that all public IPv4 and IPv6 addresses from the Internet will be able to access the hosts via ssh. The only line of defense would be the ssh key you have generated above. This is a very bad security practice and is not recommended.**

There is a workaround for this behavior, in short:

- You will need another Security Group created in the same VPC as the worker nodes. This SG will allow traffic from the bastion's SG (or it will whitelist your public IP)
- Attach this SG to the worker nodes on port 22
- Or you can always manually add a rule in the worker nodes' SG to whitelist the desired IP address on port 22

> - Note that all manual additions will prevent the eksctl stack from getting deleted, please refer to the `delete cluster` section of this guide

Reviewer: This should be treated as a stretch goal, not a requirement.

> - Note that all manual additions will prevent the eksctl stack from getting deleted, please refer to the `delete cluster` section of this guide

Author: This is not treated as a requirement, and that's why it is marked under `Optional steps`. Nothing is being added manually, it is just an explanation for the users what options are there, to save time on googling for fixes.

### Scaling a running cluster

Once the script creates a cluster, you will find a file called `$CLUSTER_NAME.yaml` in the `eks` folder you ran the script from. This is the eksctl config file for the nodegroups.
You can simply open the file and edit the field `desiredCapacity` for the `ng-2-plan` nodegroup, and add the number of nodes you wish to scale to.

Save the file and execute the following:

```
eksctl scale nodegroup --config-file=$CLUSTER_NAME.yaml --name=ng-2-plan
```

> **Please note that for the `AWS VPC CNI` scaling up and down works fine, while for `calico` you need to reset the `kube-multus-ds` daemonset and `coredns` deployment after nodegroup scaling; both located in the `kube-system` namespace. We are working on a fix that will automate this.**

Author: This was removed before you posted your comment; there is no mention of calico anywhere in the latest iteration.

This guide will help you setup an EKS (Elastic Kubernetes Service) cluster, and run testground plans on it.

## Requirements

It is assumed that you already have the following:

- A bastion host on AWS, or a laptop with AWS credentials which will be used to run the installation script, interact with the cluster, and run testground plans
- The laptop or bastion host has the following software installed:

**Note: the software and installation guides have been tested on amazon linux. Please verify which OS are you using and follow the official guides accordingly.
Links have been included for each software requirement. Listed guides are for example purposes only.**

You can find the official AWS guide for setting up a bastion host here:

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/get-set-up-for-amazon-ec2.html

**TODO: As manual setup is error-prone, and likely will get outdated pretty quickly. We need to look into creating a custom AMI.**

> **1. docker**

Author: As I agree that we do not need `docker` for the script itself, this guide is intended for users who want to setup the cluster and run tests. 
And to run tests, you need to have `docker` installed. So I think it is intuitive and good practice to list it here as a requirement.

> **1. docker**

Reviewer: We don't need docker for running tests either, do we?

> **1. docker**

Author: As far as we know, we require docker to build the images while running testground runs. 
Is there another builder that we are unaware of?

> **1. docker**

Reviewer: Yes, you require docker to be available on the cluster. That's where the build happens. That's why you were seeing the issues with goproxy, for example, on the cluster and not on your local machine.

> **1. docker**

Author: The issues with goproxy are also happening on local machines, they are not tied to EKS.
And I'm not sure what you mean by `docker to be available on the cluster`. Can you please explain this?

> **1. docker**

Reviewer 2: @galargh should we resolve this? add something to https://github.com/testground/testground/issues/1500 ?

Once the script creates a cluster, you will find a file called `$CLUSTER_NAME.yaml` in the `eks` folder you ran the script from. This is the eksctl config file for the nodegroups.
You can simply open the file and edit the field `desiredCapacity` for the `ng-2-plan` nodegroup, and add the number of nodes you wish to scale to.

Save the file and execute the following:

```
eksctl scale nodegroup --config-file=$CLUSTER_NAME.yaml --name=ng-2-plan
```

### Using a different AMI for the worker nodes

The AMI used for worker nodes defaults to Amazon EKS optimized Amazon Linux 2 v1.22 built on 08 Aug 2022, or by image name, `amazon-eks-node-1.22-v20220802`.
When you run the script, it will automatically find that image in your specified region and implement it into the config file.

If you need to specify a different image, you may use this command from your cli (and make sure you change the name to your desired AMI, and also desired Region):
```
aws ec2 describe-images --filters "Name=name,Values=amazon-eks-node-1.22-v20220802" --owners amazon --region $REGION | jq -r '.Images[0].ImageId'
```
and add a variable to the `.env` file, for example:

```
AMI_ID="ami-030609f7e95c0f386"
```

> Next step would be to edit the `make_cluster_config` function in `bash/functions.sh` and replace the `ami: $AMI_ID_REGION_BASED` with your variable.

Author: Answered [here](https://github.com/testground/infra/pull/78#discussion_r993235089).
